### PR TITLE
ch(test coverage): improve test coverage

### PR DESCRIPTION
**Pull request by ObieWalker**
What does this PR do?
This PR improves my test coverage for the backend of this application.

Description of Task to be completed?
- get test coverage above 80%
- implement edge cases
- fix pit falls arising from tests

How should this be manually tested?
* clone the repo and change into the project directory
* run npm install to install required packages
* create a .env file and add all the required environment variables
* start up the server with the command npm run test:dev

Any background context you want to provide?
N/A
What are the relevant pivotal tracker stories?(if applicable)
#157116352

### CODE REVIEW

File: server/tests/users.js [hunk 1]
```diff
@@ -0,0 +1,201 @@
+import chai from 'chai';
+import chaiHttp from 'chai-http';
+import app from '../app';
+import helper from './testHelper';
+
+const { expect } = chai;
+chai.use(chaiHttp);
+chai.should();
+
+module.exports = describe('Tests for the user', () => {
```
Comment (iidrees) [inline server/tests/users.js:10]: This is a general feedback
This is actually cool, but this does not show unit and integration testing.
your tests should as much as possible be independent of each other, meaning, no previous tests should have to depend on whether the one before it passes or fails.

So try and find out how you can write integration and unit tests because that would make your tests more robust.

File: server/tests/users.js [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+  it('should return a status 400 error response for null first name', (done) => {
```
Comment (iidrees) [inline server/tests/users.js]: I noticed that you are using `400` error `http` code when testing for user authentication (i.e. signup and signin).

I would recommend that you take look at this resource https://en.wikipedia.org/wiki/List_of_HTTP_status_codes and let it guide you on the most appropriate status codes to use for many of the responses your API endpoint returns.

File: server/tests/users.js [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+      })
+      .end((err, res) => {
+        expect(res).to.be.status(201);
+        res.body.should.have.a('object');
+        res.body.should.have.property('message').to.equal('Your account has been created!, Your details');
+        done();
+      });
+  });
+
+  it('shoud return a status 400 response the email already existing', (done) => {
```
Comment (iidrees) [inline server/tests/users.js]: Also your tests are just TDD and not BDD and your tests should be Behaviour Driven Development and not just Test Driven Development.

The importance of your test being BDD is that it clearly shows that you are testing for how user interact with your application as opposed to just writing tests to test a function or program you have written.

Also BDD helps with understanding, even non-developer who may be interested in the product understand willl understand better what it is you are trying to test for and give guidiance if you are not exactly testing based on what their preferred output is.

To help you understand better take a look at these resources 
1. https://www.upwork.com/hiring/for-clients/benefits-get-bdd/ 

2. https://www.red-gate.com/simple-talk/dotnet/net-development/behaviour-driven-development-overview-part-1-ubiquitous-language/

File: server/tests/users.js [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+      .post('/api/v1/users/login')
+      .send({
+        password: helper.userPassword,
+        email: helper.userEmail,
+      })
+      .end((err, res) => {
+        expect(res).to.be.status(200);
+        res.body.should.have.a('object');
+        res.body.should.have.property('token');
+        helper.setUserToken(res.body.token);
```
Comment (iidrees) [inline server/tests/users.js]: Just like I mentioned above, you should be able to test each ` describe` and individual `it` block without having to wait for the previous one to finish.

And by the time you have broken your code down into unit test, you will see that there would be no need to export a `token` from your `users.js` file into the global scope so that your Centers and Events  tests files can have access to them.

If you do not understand how to go about this, reach out to me and I will explain how to do this as best as I can.

File: server/tests/index.js [hunk 5]
```diff
@@ -0,0 +1,3 @@
+import users from './users';
```
Comment (iidrees) [inline server/tests/index.js:2]: I get the feeling that you did this in order to make the tests run in the order in which they were arranged.

Again, writing unit tests would not require you doing this since each test runs independent of the other tests in the whole test file.
So, look into redrafting your tests to be more unit-tests like.

File: server/tests/events.js [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+          .delete(`/api/v1/events/eventID?token=${helper.userToken}`)
+          .end((err, res) => {
+            expect(res).to.be.status(400);
+            res.should.be.a('object');
+            res.body.should.have.property('message').to.equal('There was an error with the event ID input!');
+            done();
+          });
+      });
+
+    //   it('should return a status 200 success response for deleting an event', (done) => {
```
Comment (iidrees) [inline server/tests/events.js]: I get that you may want to use these commented out  code at some other point in time, but its bad practice to push commented out codes to production.
Please if you are certain you would not be using the code, copy it out and save somewhere else or delete them.

Commented out codes are only pretty in development and not production.
So please fix this.

File: server/tests/events.js [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+        })
+        .end((err, res) => {
+          expect(res).to.be.status(404);
+          res.should.be.a('object');
+          res.body.should.have.property('message').to.equal('Event does not exist within our records');
+          done();
+        });
+    });
+
+    // it('should return a status 200 success response for valid put request', (done) => {
```
Comment (iidrees) [inline server/tests/events.js]: Same comment about comment out codes above applies here.

File: server/tests/events.js [hunk 8]
```diff
@@ -0,0 +1,399 @@
+import chai from 'chai';
+import chaiHttp from 'chai-http';
+import app from '../app';
+import helper from './testHelper';
```
Comment (iidrees) [inline server/tests/events.js:4]: if this is commented out or I go to the `user.js` test file to comment out the `export` and I run `npm test` most if not all the tests in this file will not pass.

So again, look into unit-testing

File: server/tests/events.js [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+        })
+        .end((err, res) => {
+          expect(res).to.be.status(400);
+          res.body.should.have.a('object');
+          res.body.error[0].should.have.property('msg').to.equal('Your guest estimate must be a number more than 2');
+          done();
+        });
+    });
+
+    it('should return a status 400 error for too low estimated guest', (done) => {
```
Comment (iidrees) [inline server/tests/events.js]: the comments about BDD and http status codes above also apply here.
so take note.

File: server/tests/centers.js [hunk 10]
```diff
@@ -0,0 +1,376 @@
...
+import faker from 'faker';
+import app from '../app';
+import helper from './testHelper';
+
+const { expect } = chai;
+chai.use(chaiHttp);
+chai.should();
+faker.seed(300);
+
+module.exports = describe('Tests for the centers', () => {
```
Comment (iidrees) [inline server/tests/centers.js:12]: All the comments I made above also apply here.
Finally, you can write unit tests  to ensure that your models are well tests too.

**Review by iidrees — COMMENTED** [review]
Please go through this review and start going through the comments from the `User.js` file.

File: server/tests/users.js (hunk 3 above)
Comment (ObieWalker) [reply server/tests/users.js]: Thanks a lot, I really appreciate the feedback.

File: server/tests/users.js (hunk 1 above)
Comment (ObieWalker) [reply server/tests/users.js:10]: Can we discuss more on this on Monday, I will love to clarify somethings. Thank you.
